src/App.tsx

```diff
@@ -1,24 +1,246 @@
+        setTempTodo(responce);
+        setTodosFromServer([...todosFromServer, responce]);
+      })
+      .catch(() => {
+        setError(ErrorSpec.NOT_ADDED);
+        setIsInputDisabled(false);
+      })
+      .finally(() => {
+        setIsInputDisabled(false);
+        setTempTodo(null);
+      });
+  };
+
+  const removeTodo = (id: number) => {
+    setIsInputDisabled(true);
+    setLoadingTodo(filteredTodos.find((todo) => todo.id === id) || null);
+
+    TodoService.deleteTodo(id)
+      .then(() => setLoadingTodo(null))
+      .catch(() => {
+        setError(ErrorSpec.NOT_DELETED);
+        setIsInputDisabled(false);
+      })
+      .finally(() => {
+        setTodosFromServer((prev) => prev.filter((todo) => todo.id !== id));
+        setIsInputDisabled(false);
+      });
+  };
+
+  const clearCompleted = () => {
+    filteredTodos.forEach((element) => {
+      if (element.completed) {
+        removeTodo(element.id);
+      }
+    });
+    setUncompletedTodosCount(0);
+    setHasErrors(false);
+  };
+
+  const updateTodo = (todo: Todo) => {
+    setLoadingTodo(
+      filteredTodos.find((todoToPatch) => todoToPatch.id === todo.id) || null,
+    );
+
+    TodoService.updateTodo(todo)
+      .then(() => {
+        setLoadingTodo(todo);
+      })
+      .catch(() => {
+        setError(ErrorSpec.NOT_UPDATED);
+        setLoadingTodo(null);
+      })
+      .finally(() => {
+        const todosAfterUpdate = filteredTodos.map(
+          todoFromPrev => (todoFromPrev.id === todo.id ? todo : todoFromPrev),
+        );
+
+        setTodosFromServer(todosAfterUpdate);
+        setLoadingTodo(null);
+      });
+  };
+
+  const onToggleCompleted = (todo: Todo) => {
+    const todoUpdated = { ...todo, completed: !todo.completed };
+
+    updateTodo(todoUpdated);
+  };
+
+  const onToggleAll = () => {
+    const toggledTodos = todosFromServer.map((todo) => (areAllCompleted
+      ? {
+        ...todo,
+        completed: !todo.completed,
+      }
+      : {
+        ...todo,
+        completed: true,
+      }));
+
+    const updatePromises = toggledTodos.map(
+      updatedTodo => TodoService.updateTodo(updatedTodo),
+    );
+
+    Promise.all(updatePromises)
+      .then(() => {
+        setTodosFromServer(toggledTodos);
+      })
+      .catch(() => {
+        setError(ErrorSpec.NOT_UPDATED);
+        setHasErrors(true);
+      });
+  };
+
+  const handleTitleUpdate = (newTitle: string, todo: Todo) => {
+    setEditingTodo(null);
+
+    if (!newTitle) {
+      removeTodo(todo.id);
+    } else if (newTitle !== todo.title) {
+      const updatedTodo = { ...todo, title: newTitle };
+
+      updateTodo(updatedTodo);
+    }
+  };
+
+  const onEditTodo = (todoToEdit: Todo | null) => {
+    setEditingTodo(todoToEdit);
+  };
+
+  useEffect(() => {
+    setLoadingTodo(
+      filteredTodos.find((todo) => todo.id === loadingTodo?.id) || null,
+    );
+  }, [filteredTodos, loadingTodo]);
+
   if (!USER_ID) {
+    setHasErrors(true);
```

Comment:
you can't call like that
the setter call must be in some function or event handler